File: 3748.count-stable-subarrays.py
#
# @lc app=leetcode id=3748 lang=python3
#
# [3748] Count Stable Subarrays
#

# @lc code=start
class Solution:
    def countStableSubarrays(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        results = []
        for li, ri in queries:
            count = 0
            left = li
            while left <= ri:
                right = left + 1
                # Find end of current stable segment
                while right <= ri and nums[right] >= nums[right - 1]:
                    right += 1
                # Calculate number of stable subarrays within [left, right-1]
                n = right - left
                count += n * (n + 1) // 2
                # Move left pointer to start of next potential segment
                left = right
            results.append(count)
        return results
# @lc code=end

File: 3748.count-stable-subarrays.cpp
#
# @lc app=leetcode id=3748 lang=cpp
#
# [3748] Count Stable Subarrays
#
# @lc code=start
class Solution {
public:
    vector<long long> countStableSubarrays(vector<int>& nums, vector<vector<int>>& queries) {
        vector<long long> results;
        for (const auto& query : queries) {
            int li = query[0], ri = query[1];
            long long count = 0;
            int length = 1; // Track current length of stable sequence
            for (int i = li + 1; i <= ri; ++i) {
                if (nums[i] >= nums[i - 1]) {
                    ++length;
                } else {
                    // Calculate number of stable subarrays from current sequence length
                    count += (long long)(length * (length + 1)) / 2;
                    length = 1; // Reset length for new sequence
                }
            }
            // Handle last segment if it ends with a non-decreasing sequence
            count += (long long)(length * (length + 1)) / 2;
            results.push_back(count);
        }
        return results;
    }
};
# @lc code=end